spec/unit/middleware/service_broker_rate_limiter_spec.rb

```diff
@@ -115,6 +109,80 @@ module Middleware
             end
           end
         end
+
+        shared_examples 'rate-limited endpoint' do
+          it 'does not allow more than the max number of concurrent requests' do
+            allow(app).to receive(:call) do
+              sleep(1)
+              [200, {}, 'a body']
+            end
+
+            threads = 2.times.map { Thread.new { middleware.call(env) } }
+            statuses = threads.map(&:join).map(&:value).map(&:first)
+
+            expect(statuses).to include(200)
+            expect(statuses).to include(429)
+            expect(app).to have_received(:call).once
+            expect(logger).to have_received(:info).with("Service broker concurrent rate limit exceeded for user '#{user_guid}'")
+          end
+        end
+
+        v2_rate_limited_service_endpoints = %w[/v2/service_instances /v2/service_bindings /v2/service_keys]
+
+        methods = %w[PUT POST DELETE]
+
+        v2_rate_limited_service_endpoints.each do |endpoint|
+          methods.each do |method|
+            context "#{endpoint} #{method} endpoint" do
+              let(:path) { endpoint }
+              let(:request_method) { method }
+              let(:fake_request) { instance_double(ActionDispatch::Request, fullpath: endpoint, method: method) }
+
+              include_examples 'rate-limited endpoint'
+            end
+          end
+        end
+
+        v3_rate_limited_service_endpoints = %w[/v3/service_instances/:guid/parameters /v3/service_credential_bindings/:guid/parameters /v3/service_route_bindings/:guid/parameters]
+
+        v3_rate_limited_service_endpoints.each do |endpoint|
+          context "#{endpoint} GET" do
+            let(:path) { endpoint.gsub(':guid', instance.guid.to_s) }
+            let(:request_method) { 'GET' }
+            let(:fake_request) { instance_double(ActionDispatch::Request, fullpath: path, method: request_method) }
+
+            include_examples 'rate-limited endpoint'
+          end
+        end
```

Comment:
use a dedicated context for the v3 case.
Also maybe put the whole section (shared example + v2 + v3) in a describe block.
Otherwise readability is very bad